botorch/generation/gen.py

```diff
@@ -288,19 +286,18 @@ def gen_candidates_torch(
         loss = -acquisition_function(clamped_candidates).sum()
         if verbose:
             print("Iter: {} - Value: {:.3f}".format(i, -(loss.item())))
-        loss_trajectory.append(loss.item())
-        param_trajectory["candidates"].append(clamped_candidates.clone())
 
         def closure():
             bayes_optimizer.zero_grad()
-            loss.backward()
+            output_grad = torch.autograd.grad(loss, clamped_candidates)[0]
+            clamped_candidates.grad = output_grad
             return loss
 
         bayes_optimizer.step(closure)
         with torch.no_grad():
             clamped_candidates = columnwise_clamp(
                 X=clamped_candidates, lower=lower_bounds, upper=upper_bounds
-            )
+            ).requires_grad_(True)
```

Comment:
Hmm I've never thought about what happens when you set `requires_grad` within a `no_grad` context. Will this work properly?

Reply:
Seems to be since it's not actually a differentiable operation. Wasn't passing the unit tests before.